package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,335 @@
+    >>> slope = linear_model.slope
+    >>> error = linear_model.rvalue
+    >>> D = slope * 1/(2*MSD.dim_fac) #dim_fac is 3 as we computed a 3D msd ('xyz')
+
+We have now computed a self diffusivity!
+
+
+Notes
+_____
+
+There are several factors that must be taken into account when setting up and processing trajectories for computation of self diffusivities.
+These include specific instructions around simulation settings, using unwrapped trajectories and maintaining relativley small elapsed time between saved frames. 
+Additionally corrections for finite size effects are sometimes employed along with varied means of estimating errors.
+The reader is directed to the following review, which describes many of the common pitfalls [Maginn2019]_. There are other ways to compute self diffusivity including from a Green-Kubo integral. At this point in time these methods are beyond the scope of this module
+
+
+References
+----------
+
+.. [Maginn2019] Maginn, E. J.; Messerly, R. A.; Carlson, D. J.; Roe, D. R.; Elliott, J. R. Best Practices for Computing Transport Properties 1. Self-Diffusivity and Viscosity from Equilibrium Molecular Dynamics [Article v1.0]. Living J. Comput. Mol. Sci. 2019, 1 (1).
+.. [Calandri2011] Calandrini, V.; Pellegrini, E.; Calligari, P.; Hinsen, K.; Kneller, G. R. NMoldyn-Interfacing Spectroscopic Experiments, Molecular Dynamics Simulations and Models for Time Correlation Functions. Collect. SFN 2011, 12, 201–232.
+
+Classes and Functions
+---------------------
+
+.. autoclass:: EinsteinMSD
+    :members:
+
+"""
+
+from __future__ import division, absolute_import
+from six.moves import zip
+
+import os
+import errno
+import warnings
+import bz2
+import functools
+
+import numpy as np
+from numpy.fft import fft,ifft
+import logging
+import tidynamics
+import MDAnalysis
+from ..due import due, Doi
+
+due.cite(Doi("10.21105/joss.00877"),
+         description="Mean Squared Displacements with tidynamics",
+         path="MDAnalysis.analysis.msd",
+         cite_module=True)
+due.cite(Doi("10.1051/sfn/201112010"),
+         description="FCA fast correlation algorithm",
+         path="MDAnalysis.analysis.msd",
+         cite_module=True)
+del Doi
+
+class EinsteinMSD(object):
+    r"""Class to calculate Mean Squared Displacement by the Einstein relation.
+    
+    Attributes
+    ----------
+    dim_fac : float
+        Dimensionality :math:`d` of the MSD.
+    timeseries : :class:`np.ndarray`
+        The averaged MSD with respect to lag-time.
+    msd_per_particle : :class:`np.ndarray`
+        The MSD of each individual particle with respect to lag-time.
+    N_frames : int
+        Number of frames in trajectory.
+    N_particles : int
+        Number of particles MSD was calculated over.
+
+    """
+
+    def __init__(self, u, selection=None, msd_type='xyz', fft=True):
+        r"""
+        Parameters
+        ----------
+        u : Universe
+            An MDAnalysis :class:`Universe`.
+        selection : str
```

Comment:
I think it would be good for future-proofing to design so that the `u` keyword can accept either an `AtomGroup` or `Universe`; it's usually easily achievable as `u.universe` and `AtomGroup.universe` are interchangeable, as are `AtomGroup.atoms` and `Universe.atoms`. I think @RMeli has a couple PR's devoted to moving `selection` and other keywords to the standard AnalysisBase `select` argument, although I think `select` has conventionally been strings only? #2528 and #2531 were to this effect.

The actual attempt to make `AtomGroup`s interchangeable with selection strings ended in [prudent postponement](https://github.com/MDAnalysis/mdanalysis/pull/2375#issuecomment-547187106) due to complexities with `UpdatingAtomGroup`.

Reply:
This is insightful, but I am a little confused about what course of action to take. Given that they are not interchangeable are we staying with a string?

Reply:
Name the kwarg `select` (see #2461), make it accept an `AtomGroup` or a `Universe`, don't bother with strings.

Make sure that it does whatever it has to do with an [`UpdatingAtomGroup`](https://www.mdanalysis.org/UserGuide/selections.html#dynamic-selections): if it cannot deal with it, raise a `TypeError`. Otherwise make sure (test!) that it works, i.e., when you get varying number of coordinates for each frame.

Reply:
Perhaps make `u` accept an `AtomGroup` or `Universe` (and while you're at it, rename it to `atomgroup` or `universe` to bring it more in line with newer analysis classes)? `select` is conventionally a string that is used to select the group from the given universe/atomgroup as [`select_atoms()` cannot accept an AtomGroup](https://github.com/MDAnalysis/mdanalysis/issues/2369).

Reply:
@lilyminium we have a bunch of classes that happily use `select` for an AtomGroup and not a string. I had a look around before I wrote my comment because I initially wondered if we should reserve `select` for strings. But I think at this stage all that we can say for 1.0 is that `select` is a kind of selection and the class decides if it wants a string or an AG. In 2.0 we will make `select` exclusively accept AGs.

This is a bit ugly in terms of UI; the alternative would be to back-introduce `selection` for AG and `select` for strings.

Reply:
(Sorry @hmacdope ... apparently we hadn't really settled on how to handle `select`.)

Reply:
@lilyminium , you make a decision for this PR, you know at least as much as I do about the subject. 

(I've got to finish parts of a paper... or maybe actually catch some sleep.)

Reply:
Hope the paper doesn't take too long and you do get some sleep, @orbeckst.

Going through all the analysis classes, 

* the leaflet module accepts AtomGroups or strings for `select`, although it's not an optional argument
* LinearDensity only accepts an `AtomGroup` for `select`, although given that it's the first argument and non-optional it could be argued it's essentially the same argument as `atomgroup`/`universe` in other classes
* some other classes (eg InterRDF) only accept `AtomGroup`s and not strings, but don't call these arguments `select`.

(Do correct me if I'm wrong or missed one, @RMeli.)

As every single other analysis class with the keyword `select` only accepts strings or tuples of strings, let's stick with a string for this PR and sort out better selection methods later.

Reply:
> This is a bit ugly in terms of UI; the alternative would be to back-introduce selection for AG and select for strings.

@orbeckst I'm not a big fan of this because of what already discussed in #2461. I think that going forward we should add type hints (see discussion in #2450) so that it will become clear if `select` takes a `str`, a `tuple` or an `AtomGroup` (and incorrect uses can be spotted without running the code by using `mypy`).

> As every other analysis class with the keyword select only accepts strings or tuples of strings, let's stick with a string for this PR and sort out better selection methods later.

@lilyminium I don't recall which classes accepted `AtomGroup`s or `str` but if it's only the few you listed your suggestion sounds very reasonable for the time being.

Reply:
In theory I like type-hinting. In practice this feels long (a `select` keyword that accepts a string, AtomGroup, or dict or Tuple of either, e.g. for rms.RMSD):

```python
select: Union[str, AtomGroup, Dict[str, Union[str, AtomGroup]], Tuple[Union[str, AtomGroup], Union[str, AtomGroup]]]="all"
```

but that's a discussion to be had elsewhere.

Reply:
I agree, but we could use [type aliases](https://docs.python.org/3/library/typing.html#type-aliases):
```
SelectTypes = Union[str, AtomGroup, Dict[str, Union[str, AtomGroup]], Tuple[Union[str, AtomGroup], Union[str, AtomGroup]]]

select: SelectTypes = "all"
```

This will not make the types explicit but it will already hint that there are multiple accepted types, and `mypy` can work them out. 

Please comment on #2450, it's definitely something to be discussed.